File: axiom/lib/services/data_insights_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

class DataInsightsService {
  static const String baseUrl = 'http://localhost:5000/api/data-insights';

  static Future<FormCompletionInsights> getFormCompletionInsights({
    required String projectId,
    required String formId,
  }) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/form-completion'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({'projectId': projectId, 'formId': formId}),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        if (data['success']) {
          return FormCompletionInsights.fromJson(data['data']);
        } else {
          throw Exception(data['error'] ?? 'Failed to get form completion insights');
        }
      }
      throw Exception('Failed to get form completion insights');
    } catch (e) {
      throw Exception('Data Insights Service Error: $e');
    }
  }

  static Future<DataPatternRecognition> getDataPatternRecognition({
    required String projectId,
    required String dataType,
  }) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/data-patterns'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({'projectId': projectId, 'dataType': dataType}),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        if (data['success']) {
          return DataPatternRecognition.fromJson(data['data']);
        } else {
          throw Exception(data['error'] ?? 'Failed to analyze data patterns');
        }
      }
      throw Exception('Failed to analyze data patterns');
    } catch (e) {
      throw Exception('Data Insights Service Error: $e');
    }
  }

  static Future<PerformanceRecommendations> getPerformanceRecommendations({
    required String projectId,
    List<Map<String, dynamic>>? apiUsage,
  }) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/performance-recommendations'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({
          'projectId': projectId,
          'apiUsage': apiUsage ?? [],
        }),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        if (data['success']) {
          return PerformanceRecommendations.fromJson(data['data']);
        } else {
          throw Exception(data['error'] ?? 'Failed to get performance recommendations');
        }
      }
      throw Exception('Failed to get performance recommendations');
    } catch (e) {
      throw Exception('Data Insights Service Error: $e');
    }
  }
}

// Data models
class FormCompletionInsights {
  final double overallCompletionRate;
  final List<DropOffPoint> dropOffPoints;
  final List<Recommendation> recommendations;

  FormCompletionInsights({
    required this.overallCompletionRate,
    required this.dropOffPoints,
    required this.recommendations,
  });

  factory FormCompletionInsights.fromJson(Map<String, dynamic> json) {
    return FormCompletionInsights(
      overallCompletionRate: (json['completionRate']['overall'] ?? 0.0).toDouble(),
      dropOffPoints: (json['dropOffPoints'] as List?)
          ?.map((d) => DropOffPoint.fromJson(d))
          .toList() ?? [],
      recommendations: (json['recommendations'] as List?)
          ?.map((r) => Recommendation.fromJson(r))
          .toList() ?? [],
    );
  }
}

class DropOffPoint {
  final String fieldName;
  final double dropOffRate;
  final String severity;
  final List<String> suggestions;

  DropOffPoint({
    required this.fieldName,
    required this.dropOffRate,
    required this.severity,
    required this.suggestions,
  });

  factory DropOffPoint.fromJson(Map<String, dynamic> json) {
    return DropOffPoint(
      fieldName: json['fieldName'] ?? '',
      dropOffRate: (json['dropOffRate'] ?? 0.0).toDouble(),
      severity: json['severity'] ?? 'low',
      suggestions: List<String>.from(json['suggestions'] ?? []),
    );
  }
}

class Recommendation {
  final String type;
  final String priority;
  final String title;
  final String description;

  Recommendation({
    required this.type,
    required this.priority,
    required this.title,
    required this.description,
  });

  factory Recommendation.fromJson(Map<String, dynamic> json) {
    return Recommendation(
      type: json['type'] ?? '',
      priority: json['priority'] ?? 'medium',
      title: json['title'] ?? '',
      description: json['description'] ?? '',
    );
  }
}

class DataPatternRecognition {
  final List<DataPattern> detectedPatterns;
  final List<FieldTypeSuggestion> suggestedFieldTypes;

  DataPatternRecognition({
    required this.detectedPatterns,
    required this.suggestedFieldTypes,
  });

  factory DataPatternRecognition.fromJson(Map<String, dynamic> json) {
    return DataPatternRecognition(
      detectedPatterns: (json['detectedPatterns'] as List?)
          ?.map((p) => DataPattern.fromJson(p))
          .toList() ?? [],
      suggestedFieldTypes: (json['suggestedFieldTypes'] as List?)
          ?.map((s) => FieldTypeSuggestion.fromJson(s))
          .toList() ?? [],
    );
  }
}

class DataPattern {
  final String patternType;
  final double confidence;
  final List<String> examples;

  DataPattern({
    required this.patternType,
    required this.confidence,
    required this.examples,
  });

  factory DataPattern.fromJson(Map<String, dynamic> json) {
    return DataPattern(
      patternType: json['patternType'] ?? '',
      confidence: (json['confidence'] ?? 0.0).toDouble(),
      examples: List<String>.from(json['examples'] ?? []),
    );
  }
}

class FieldTypeSuggestion {
  final String fieldName;
  final String suggestedType;
  final double confidence;

  FieldTypeSuggestion({
    required this.fieldName,
    required this.suggestedType,
    required this.confidence,
  });

  factory FieldTypeSuggestion.fromJson(Map<String, dynamic> json) {
    return FieldTypeSuggestion(
      fieldName: json['fieldName'] ?? '',
      suggestedType: json['suggestedType'] ?? '',
      confidence: (json['confidence'] ?? 0.0).toDouble(),
    );
  }
}

class PerformanceRecommendations {
  final Map<String, dynamic> apiOptimizations;
  final Map<String, dynamic> cachingStrategies;
  final Map<String, dynamic> databaseOptimizations;
  final Map<String, dynamic> frontendOptimizations;

  PerformanceRecommendations({
    required this.apiOptimizations,
    required this.cachingStrategies,
    required this.databaseOptimizations,
    required this.frontendOptimizations,
  });

  factory PerformanceRecommendations.fromJson(Map<String, dynamic> json) {
    return PerformanceRecommendations(
      apiOptimizations: json['apiOptimizations'] ?? {},
      cachingStrategies: json['cachingStrategies'] ?? {},
      databaseOptimizations: json['databaseOptimizations'] ?? {},
      frontendOptimizations: json['frontendOptimizations'] ?? {},
    );
  }
}

File: axiomBackend/src/services/dataInsightsService.js
class DataInsightsService {
  async getFormCompletionInsights(projectId, formId) {
    console.log('📊 Analyzing form completion insights for project:', projectId);
    
    // Simulate AI processing time
    await new Promise(resolve => setTimeout(resolve, 1000));
    
    // Mock form completion data
    const mockData = this.generateMockCompletionData(formId);
    
    const insights = {
      completionRate: this.calculateCompletionRate(mockData),
      dropOffPoints: this.identifyDropOffPoints(mockData),
      averageTimeToComplete: this.calculateAverageCompletionTime(mockData),
      fieldAnalytics: this.analyzeFieldPerformance(mockData),
      recommendations: this.generateCompletionRecommendations(mockData),
      trends: this.analyzeCompletionTrends(mockData)
    };
    
    return insights;
  }

  async getDataPatternRecognition(projectId, dataType) {
    console.log('🔍 Recognizing data patterns for:', dataType);
    
    await new Promise(resolve => setTimeout(resolve, 800));
    
    const patterns = {
      detectedPatterns: this.identifyDataPatterns(dataType),
      suggestedFieldTypes: this.suggestFieldTypes(dataType),
      dataValidation: this.recommendValidation(dataType),
      structureSuggestions: this.analyzeDataStructure(dataType),
      optimizationTips: this.generateDataOptimizationTips(dataType)
    };
    
    return patterns;
  }

  async getPerformanceRecommendations(projectId, apiUsage) {
    console.log('⚡ Analyzing performance for project:', projectId);
    
    await new Promise(resolve => setTimeout(resolve, 600));
    
    const recommendations = {
      apiOptimizations: this.analyzeAPIPerformance(apiUsage),
      cachingStrategies: this.recommendCaching(apiUsage),
      databaseOptimizations: this.suggestDatabaseImprovements(apiUsage),
      frontendOptimizations: this.recommendFrontendImprovements(apiUsage),
      securityRecommendations: this.analyzeSecurityPerformance(apiUsage),
      costOptimizations: this.suggestCostReductions(apiUsage)
    };
    
    return recommendations;
  }

  generateMockCompletionData(formId) {
    return {
      formId: formId,
      totalSubmissions: 1250,
      completedSubmissions: 980,
      partialSubmissions: 270,
      averageTimeSpent: 245, // seconds
      fieldData: [
        {
          fieldName: 'name',
          completionRate: 0.95,
          averageTimeToFill: 8.5,
          dropOffRate: 0.05,
          errors: 12
        },
        {
          fieldName: 'email',
          completionRate: 0.88,
          averageTimeToFill: 12.3,
          dropOffRate: 0.12,
          errors: 45
        },
        {
          fieldName: 'phone',
          completionRate: 0.72,
          averageTimeToFill: 15.7,
          dropOffRate: 0.28,
          errors: 89
        },
        {
          fieldName: 'address',
          completionRate: 0.65,
          averageTimeToFill: 22.1,
          dropOffRate: 0.35,
          errors: 67
        },
        {
          fieldName: 'message',
          completionRate: 0.58,
          averageTimeToFill: 45.2,
          dropOffRate: 0.42,
          errors: 23
        }
      ]
    };
  }

  calculateCompletionRate(mockData) {
    return {
      overall: mockData.completedSubmissions / mockData.totalSubmissions,
      byField: mockData.fieldData.map(field => ({
        fieldName: field.fieldName,
        rate: field.completionRate
      })),
      trend: { direction: 'increasing', percentage: 12.5 }
    };
  }

  identifyDropOffPoints(mockData) {
    return mockData.fieldData
      .filter(field => field.dropOffRate > 0.2)
      .map(field => ({
        fieldName: field.fieldName,
        dropOffRate: field.dropOffRate,
        position: mockData.fieldData.indexOf(field),
        severity: field.dropOffRate > 0.4 ? 'high' : field.dropOffRate > 0.3 ? 'medium' : 'low',
        suggestions: this.generateDropOffSuggestions(field)
      }))
      .sort((a, b) => b.dropOffRate - a.dropOffRate);
  }

  generateDropOffSuggestions(field) {
    if (field.fieldName === 'phone') {
      return ['Consider making phone optional', 'Add phone format helper', 'Use country code dropdown'];
    }
    if (field.fieldName === 'address') {
      return ['Use autocomplete for addresses', 'Split address into multiple fields', 'Make address optional'];
    }
    return ['Review field necessity', 'Add helper text', 'Improve field layout'];
  }

  generateCompletionRecommendations(mockData) {
    const recommendations = [];
    
    // High drop-off fields
    const highDropOffFields = mockData.fieldData.filter(f => f.dropOffRate > 0.3);
    if (highDropOffFields.length > 0) {
      recommendations.push({
        type: 'drop_off',
        priority: 'high',
        title: 'Reduce Form Abandonment',
        description: `${highDropOffFields.length} fields have high drop-off rates`,
        actions: highDropOffFields.map(field => ({
          field: field.fieldName,
          suggestion: this.getDropOffSuggestion(field),
          impact: Math.round(field.dropOffRate * 100 * 0.8)
        }))
      });
    }
    
    return recommendations;
  }

  getDropOffSuggestion(field) {
    const suggestions = {
      'phone': 'Add phone format validation and helper text',
      'address': 'Use Google Places autocomplete',
      'message': 'Add character count and rich text editor',
      'email': 'Add email validation and domain suggestions'
    };
    return suggestions[field.fieldName] || 'Review field design and validation';
  }

  identifyDataPatterns(dataType) {
    return [
      {
        patternType: 'emailPatterns',
        confidence: 0.95,
        examples: ['user@example.com', 'john.doe@company.org'],
        characteristics: ['@ symbol', 'Domain extension', 'Standard format']
      },
      {
        patternType: 'phonePatterns',
        confidence: 0.80,
        examples: ['(555) 123-4567', '555-123-4567', '+1 555 123 4567'],
        characteristics: ['Phone format', 'Area codes', 'Various separators']
      }
    ];
  }

  suggestFieldTypes(dataType) {
    const patterns = this.identifyDataPatterns(dataType);
    
    return patterns.map(pattern => ({
      fieldName: pattern.patternType,
      suggestedType: this.mapPatternToFieldType(pattern.patternType),
      confidence: pattern.confidence,
      validation: this.suggestValidation(pattern),
      examples: pattern.examples
    }));
  }

  mapPatternToFieldType(patternType) {
    const mapping = {
      'emailPatterns': 'email',
      'phonePatterns': 'phone',
      'datePatterns': 'date',
      'numericPatterns': 'number',
      'stringPatterns': 'text'
    };
    return mapping[patternType] || 'text';
  }

  suggestValidation(pattern) {
    const validations = {
      'emailPatterns': ['email format', 'domain validation'],
      'phonePatterns': ['phone format', 'country code'],
      'datePatterns': ['date format', 'range validation']
    };
    return validations[pattern.patternType] || ['required field'];
  }

  analyzeAPIPerformance(apiUsage) {
    return {
      slowEndpoints: apiUsage.filter(api => api.averageResponseTime > 1000),
      highErrorRate: apiUsage.filter(api => api.errorRate > 0.05),
      recommendations: apiUsage.map(api => ({
        endpoint: api.endpoint,
        optimization: this.suggestAPIOptimization(api),
        potentialImprovement: this.calculatePotentialImprovement(api)
      }))
    };
  }

  suggestAPIOptimization(api) {
    if (api.averageResponseTime > 2000) {
      return 'Add database indexing and query optimization';
    }
    if (api.callCount > 5000) {
      return 'Implement caching and response compression';
    }
    return 'Review query efficiency and add pagination';
  }

  calculatePotentialImprovement(api) {
    if (api.averageResponseTime > 2000) {
      return Math.round((api.averageResponseTime - 500) / api.averageResponseTime * 100);
    }
    return 15; // Default improvement
  }

  analyzeFieldPerformance(mockData) {
    return mockData.fieldData.map(field => ({
      fieldName: field.fieldName,
      performance: {
        efficiency: Math.min(100, (1 - field.dropOffRate) * 100),
        userSatisfaction: Math.max(0, 100 - (field.errors / 10) - (field.averageTimeToFill / 5)),
        errorRate: field.errors / mockData.totalSubmissions,
        completionSpeed: Math.max(0, 100 - (field.averageTimeToFill / 50) * 100)
      },
      recommendations: this.generateFieldRecommendations(field)
    }));
  }

  generateFieldRecommendations(field) {
    const recommendations = [];
    
    if (field.dropOffRate > 0.3) {
      recommendations.push('Reduce field complexity or make optional');
    }
    
    if (field.averageTimeToFill > 20) {
      recommendations.push('Add autocomplete or simplify input');
    }
    
    if (field.errors > 20) {
      recommendations.push('Add better validation and helper text');
    }
    
    return recommendations;
  }

  calculateAverageCompletionTime(mockData) {
    return {
      overall: mockData.averageTimeSpent,
      byField: mockData.fieldData.map(field => ({
        fieldName: field.fieldName,
        averageTime: field.averageTimeToFill,
        relativeTime: field.averageTimeToFill / mockData.averageTimeSpent
      })),
      benchmarks: { excellent: 10, good: 15, average: 20, poor: 30 }
    };
  }

  analyzeCompletionTrends(mockData) {
    return {
      timeSeriesData: this.generateTimeSeriesData(),
      userSegments: [
        { segment: 'Desktop', completionRate: 0.82, averageTime: 220 },
        { segment: 'Mobile', completionRate: 0.68, averageTime: 280 },
        { segment: 'Tablet', completionRate: 0.75, averageTime: 240 }
      ]
    };
  }

  generateTimeSeriesData() {
    const data = [];
    const now = Date.now();
    for (let i = 30; i >= 0; i--) {
      const date = new Date(now - (i * 24 * 60 * 60 * 1000));
      data.push({
        date: date.toISOString().split('T')[0],
        submissions: Math.floor(Math.random() * 50) + 20,
        completions: Math.floor(Math.random() * 40) + 15,
        averageTime: Math.floor(Math.random() * 100) + 180
      });
    }
    return data;
  }

  recommendValidation(dataType) {
    const patterns = this.identifyDataPatterns(dataType);
    
    return patterns.map(pattern => ({
      fieldType: pattern.patternType,
      validationRules: this.generateValidationRules(pattern),
      errorMessages: this.generateErrorMessages(pattern),
      helperText: this.generateHelperText(pattern)
    }));
  }

  generateValidationRules(pattern) {
    return {
      required: true,
      minLength: pattern.patternType === 'phonePatterns' ? 10 : 2,
      maxLength: pattern.patternType === 'addressPatterns' ? 200 : 100,
      pattern: this.getValidationPattern(pattern.patternType)
    };
  }

  getValidationPattern(patternType) {
    const patterns = {
      'emailPatterns': '/^[\\w-\\.]+@([\\w-]+\\.)+[\\w-]{2,4}$/',
      'phonePatterns': '/^\\+?[\\d\\s\\-\\(\\)]{10,}$/',
      'datePatterns': '/^\\d{4}-\\d{2}-\\d{2}$/',
      'numericPatterns': '/^\\d+$/',
      'stringPatterns': '/^[a-zA-Z\\s]+$/'
    };
    return patterns[patternType] || '/.+/';
  }

  generateErrorMessages(pattern) {
    return {
      required: 'This field is required',
      invalid: 'Please enter a valid ' + pattern.patternType,
      minLength: 'Minimum length is ' + (pattern.patternType === 'phonePatterns' ? 10 : 2) + ' characters',
      maxLength: 'Maximum length is ' + (pattern.patternType === 'addressPatterns' ? 200 : 100) + ' characters'
    };
  }

  generateHelperText(pattern) {
    const helpers = {
      'emailPatterns': 'Enter your email address (e.g., user@example.com)',
      'phonePatterns': 'Enter your phone number with area code',
      'datePatterns': 'Enter date in YYYY-MM-DD format',
      'numericPatterns': 'Enter numbers only',
      'stringPatterns': 'Enter text characters only',
      'addressPatterns': 'Enter your full address'
    };
    return helpers[pattern.patternType] || 'Enter valid information';
  }

  recommendCaching(apiUsage) {
    return {
      cacheableEndpoints: apiUsage.filter(api => this.isCacheable(api)),
      cachingStrategies: apiUsage.map(api => ({
        endpoint: api.endpoint,
        strategy: this.recommendCacheStrategy(api),
        ttl: this.recommendTTL(api),
        estimatedImprovement: this.estimateCacheImprovement(api)
      })),
      implementation: this.generateCacheImplementation()
    };
  }

  isCacheable(api) {
    return api.method === 'GET' && api.averageResponseTime > 200;
  }

  recommendCacheStrategy(api) {
    if (api.dataSize > 1000000) {
      return 'Redis with compression';
    }
    return 'Memory cache with 5-minute TTL';
  }

  recommendTTL(api) {
    if (api.frequency === 'high') {
      return 300; // 5 minutes
    }
    return 1800; // 30 minutes
  }

  estimateCacheImprovement(api) {
    return Math.round(api.averageResponseTime * 0.8); // 80% improvement
  }

  generateCacheImplementation() {
    return {
      technology: 'Redis',
      configuration: {
        maxMemory: '256MB',
        evictionPolicy: 'LRU',
        compression: true
      },
      integration: 'Add caching middleware to Express routes'
    };
  }

  suggestDatabaseImprovements(apiUsage) {
    return {
      indexing: [
        { field: 'created_at', type: 'btree' },
        { field: 'user_id', type: 'hash' },
        { field: 'status', type: 'btree' }
      ],
      queryOptimization: {
        addPagination: 'Implement LIMIT and OFFSET',
        selectFields: 'Use specific field selection',
        joinOptimization: 'Optimize JOIN queries'
      },
      connectionPooling: {
        minConnections: 5,
        maxConnections: 20,
        idleTimeout: 30000
      }
    };
  }

  recommendFrontendImprovements(apiUsage) {
    return {
      lazyLoading: {
        components: ['images', 'data tables', 'charts'],
        implementation: 'React.lazy() and Suspense'
      },
      codeSplitting: {
        strategy: 'Route-based code splitting',
        tools: ['Webpack', 'React.lazy'],
        expectedReduction: '40% bundle size'
      },
      imageOptimization: {
        formats: ['WebP', 'AVIF'],
        compression: '80% quality',
        lazyLoading: true,
        cdn: 'Use CDN for static assets'
      }
    };
  }

  analyzeSecurityPerformance(apiUsage) {
    return {
      authentication: 'Implement JWT with refresh tokens',
      authorization: 'Add role-based access control',
      dataEncryption: 'Encrypt sensitive data at rest',
      rateLimiting: 'Implement API rate limiting'
    };
  }

  suggestCostReductions(apiUsage) {
    return {
      serverOptimization: 'Right-size server instances',
      cdnUsage: 'Optimize CDN caching strategies',
      databaseOptimization: 'Use read replicas for read-heavy operations'
    };
  }

  analyzeDataStructure(dataType) {
    return {
      complexity: 'medium',
      relationships: ['one-to-many', 'many-to-many'],
      normalization: 'Third normal form recommended',
      indexing: 'Add composite indexes for frequent queries'
    };
  }

  generateDataOptimizationTips(dataType) {
    return [
      'Use appropriate data types for storage efficiency',
      'Implement data validation at database level',
      'Consider data archiving for old records',
      'Use connection pooling for better performance'
    ];
  }
}

export default new DataInsightsService();
